### PR TITLE
[Merged by Bors] - feat(Analysis/Analytic/Meromorphic): more API for meromorphic functions

### PR DESCRIPTION
**Pull request by loefflerd**
This PR adds "order of vanishing" for meromorphic functions, and some basic API for functions `MeromorphicOn` a set.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Analytic/Meromorphic.lean [hunk 1]
```diff
@@ -122,4 +131,144 @@ lemma div {f g : 𝕜 → 𝕜} {x : 𝕜} (hf : MeromorphicAt f x) (hg : Meromo
...
+    replace hg_eq : ∀ᶠ (z : 𝕜) in 𝓝[≠] x, f z = (z - x) ^ (↑m - ↑hf.choose : ℤ) • g z
+    · rw [eventually_nhdsWithin_iff]
+      filter_upwards [hg_eq] with z hg_eq hz
+      rwa [← smul_right_inj <| zpow_ne_zero _ (sub_ne_zero.mpr hz), ← mul_smul,
+      ← zpow_add₀ (sub_ne_zero.mpr hz), ← add_sub_assoc, add_sub_cancel', zpow_ofNat, zpow_ofNat]
+    exact ⟨fun h ↦ ⟨g, hg_an, hg_ne, h ▸ hg_eq⟩,
+      AnalyticAt.unique_eventuallyEq_zpow_smul_nonzero ⟨g, hg_an, hg_ne, hg_eq⟩⟩
+
+/-- Compatibility of notions of `order` for analytic and meromorphic functions. -/
+lemma order_of_analyticAt {f : 𝕜 → E} {x : 𝕜} (hf : AnalyticAt 𝕜 f x) :
```
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Meromorphic.lean:196]: could you change the name of the lemma and move it into the `AnalyticAt` namespace?

File: Mathlib/Analysis/Analytic/Meromorphic.lean [hunk 2]
```diff
@@ -122,4 +131,144 @@ lemma div {f g : 𝕜 → 𝕜} {x : 𝕜} (hf : MeromorphicAt f x) (hg : Meromo
...
+  · obtain ⟨n, hn⟩ := WithTop.ne_top_iff_exists.mp ho
+    simp_rw [← hn, WithTop.map_coe, order_eq_int_iff, zpow_ofNat]
+    rcases (hf.order_eq_nat_iff _).mp hn.symm with ⟨g, h1, h2, h3⟩
+    exact ⟨g, h1, h2, h3.filter_mono nhdsWithin_le_nhds⟩
+
+lemma iff_eventuallyEq_zpow_smul_analyticAt {f : 𝕜 → E} {x : 𝕜} : MeromorphicAt f x ↔
+    ∃ (n : ℤ) (g : 𝕜 → E), AnalyticAt 𝕜 g x ∧ ∀ᶠ z in 𝓝[≠] x, f z = (z - x) ^ n • g z := by
+  constructor
+  · intro hf_m
+    rcases eq_or_ne hf_m.order ⊤ with ho | ho
```
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Meromorphic.lean:210]: do you really need to use the order for this? You could just start from the definition of a meromorphic function, and multiply by `(z - x) ^ (-n)`.

File: Mathlib/Analysis/Analytic/Meromorphic.lean [hunk 3]
```diff
@@ -122,4 +131,144 @@ lemma div {f g : 𝕜 → 𝕜} {x : 𝕜} (hf : MeromorphicAt f x) (hg : Meromo
...
 end MeromorphicAt
+
+/-- Meromorphy of a function on a set. -/
+def MeromorphicOn (f : 𝕜 → E) (U : Set 𝕜) : Prop := ∀ x ∈ U, MeromorphicAt f x
+
+lemma AnalyticOn.meromorphicOn {f : 𝕜 → E} {U : Set 𝕜} (hf : AnalyticOn 𝕜 f U) :
+    MeromorphicOn f U :=
+  fun x hx ↦ (hf x hx).meromorphicAt
+
+lemma meromorphicOn_id {U : Set 𝕜} : MeromorphicOn id U := fun x _ ↦ meromorphicAt_id x
```
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Meromorphic.lean:229]: With the new dot notation in Lean 4, this kind of lemma should be named `MeromorphicOn.id`. The reason is that, if you're proving that a function is `MeromorphicOn`, then Lean 4 will let you use `.id` for `MeromorphicOn.id`, and it turns out to be really convenient. Same thing below for `const`.

**Review by sgouezel — COMMENTED** [review]
bors d+
Thanks!

**Comment by loefflerd** [comment]
Many thanks @sgouezel for the review!

bors r+

**Pull request closed without merging**